c++/src/kj/string.c++

```diff
@@ -489,6 +461,76 @@ char* FloatToBuffer(float value, char* buffer) {
   return buffer;
 }
 
+// ----------------------------------------------------------------------
+// NoLocaleStrtod()
+//   This code will make you cry.
+// ----------------------------------------------------------------------
+
+namespace {
+
+// Returns a string identical to *input except that the character pointed to
+// by radix_pos (which should be '.') is replaced with the locale-specific
+// radix character.
+kj::String LocalizeRadix(const char* input, const char* radix_pos) {
+  // Determine the locale-specific radix character by calling sprintf() to
+  // print the number 1.5, then stripping off the digits.  As far as I can
+  // tell, this is the only portable, thread-safe way to get the C library
+  // to divuldge the locale's radix character.  No, localeconv() is NOT
+  // thread-safe.
+  char temp[16];
+  int size = sprintf(temp, "%.1f", 1.5);
+  KJ_ASSERT(temp[0] == '1');
+  KJ_ASSERT(temp[size-1] == '5');
+  KJ_ASSERT(size <= 6);
+
+  // Now replace the '.' in the input with it.
+  return kj::str(
+      kj::arrayPtr(input, radix_pos),
+      kj::arrayPtr(temp + 1, size - 2),
+      kj::StringPtr(radix_pos + 1));
+}
+
+}  // namespace
+
+double NoLocaleStrtod(const char* text, char** original_endptr) {
+  // We cannot simply set the locale to "C" temporarily with setlocale()
+  // as this is not thread-safe.  Instead, we try to parse in the current
+  // locale first.  If parsing stops at a '.' character, then this is a
+  // pretty good hint that we're actually in some other locale in which
+  // '.' is not the radix character.
+
+  char* temp_endptr;
+  double result = strtod(text, &temp_endptr);
```

Comment:
Hmm.... what if we leverage https://github.com/lemire/fast_double_parser? Then this parsing would be ~16x faster (8x faster than strtod & you call strtod twice, + there's a sprintf in there) & simpler to reason about, although I'm sure it's correct since you're just copying your old protobuf implementation. It's also only a single header file so there's not even any build files that would change...

Reply:
Also, even if we don't take it, reading the source there, I like it's locale solution much better than what we have here using strtod_l for most systems except cygwin & solaris where it has a different solution. Unclear which is faster in that path - I suspect they still win because a small malloc seems faster than doing 2*strtod + sprintf.

Reply:
Oh, and reading the code, it looks like they technically do have a memory leak on exit that valgrind/asan might complain about (would need to change the code so that the locale is freed on shutdown)

Reply:
The code I copied from protobuf is code I wrote. :)

But I wrote it ~15 years ago. Absurdly enough, there have been advances in both floating-point parsing and stringification since then. I think it would make a lot of sense to investigate the state of the art and possibly bring in some better implementations here. But, my goal for this PR was to fix the bugs as quickly as possible in order to do a release... we can investigate these options another time.